dynamic_programming/longest_common_subsequence.cpp

```diff
@@ -0,0 +1,287 @@
+			index++;
+			i--;
+			j--;
+		}
+		// If not matched
+		else
+		{	
+			if(c[i-1][j] > c[i][j-1])
+			{
+				i--;
+			}
+			else if(c[i-1][j] == c[i][j-1])
+			{
+                // string will contain one of the LCS
+                char *string = allPossibleLCS(c, a, index, i, j-1, string1, string2);
+
+                // if LCSArray is NULL then allocating a memory to store LCS
+                if(LCSArray == NULL)
+                {
+                    LCSArray = (char **)calloc(1, sizeof(char *));
+                    *(LCSArray + ind) = (char *)calloc(lengthOfLCS, sizeof(char));
+                    for(int k=lengthOfLCS-1; k >= 0; k--)
+                    {
+                        LCSArray[ind][lengthOfLCS-k-1] = string[k];
+                    }
+                }
+                // if not NULL then reallocating to store one more LCS
+                else
+                {
+                    LCSArray = (char **)realloc(LCSArray, (ind + 2)*sizeof(char *));
+                    ind++;
+                    *(LCSArray + ind) = (char *)calloc(lengthOfLCS, sizeof(char));
+                    for(int k=lengthOfLCS-1; k >= 0; k--)
+                    {
+                        LCSArray[ind][lengthOfLCS-k-1] = string[k];
+                    }
+                }
+				i--;
+			}
+			else
+			{
+				j--;
+			}
+		}
+	}
+	return a;
+}
+
+/**
+ * @brief max will find the maximum between 2 values
+ * @param a integer a
+ * @param b integer b
+ * @returns maximum value
+ */
+int max(int a, int b)
+{
+	// if a is greater than b return a
+	if(a > b)
+	{
+		return a;
+	}
+	// else return b
+	return b;
+}
+
+/**
+ * @brief LCS will make 2D-array with optimal solutions of sub-problems
+ * @param string1 String 1
+ * @param m Length of String 1
+ * @param string2 String 2
+ * @param n Length of String 2
+ * @returns 2D-array
+ */
+int** LCS(char *string1,int m, char *string2, int n)
+{
+	// Creating a 2D-array to store the solutions of all sub-problems
+	int **c;
+	
+	// Dynamically allocating memory for 2D-array
+	c = (int **)calloc(m+1, sizeof(int *));
+	for(int i=0; i<=m; i++)
+	{
+		*(c + i) = (int *)calloc(n+1, sizeof(int));
+	}
+	// c[i][j] will have the length of Longest common subsequence when length of string 1 is i and length of string 2 is j
+	
+	// If length of string 2 is 0 then there will be no Longest common subsequence
+	for(int i=1; i<=m; i++)
+	{
+		c[i][0] = 0;
+	}
+	
+	// If length of string 1 is 0 then there will be no Longest common subsequence
+	for(int j=0; j <=n; j++)
+	{
+		c[0][j] = 0;
+	}
+	
+	for(int i=0; i<m; i++)
+	{
+		for(int j=0; j<n; j++)
+		{
+			// If string1[i] == string2[j]
+			if(string1[i] == string2[j])
+			{
+				c[i+1][j+1] = c[i][j]+1;
+			}
+			// If string1[i] != string2[j]
+			else
+			{
+				c[i+1][j+1] = max(c[i][j+1], c[i+1][j]);
+			}
+		}
+	}
+	return c;
+}
+
+/**
+ * @brief test is to test the algorithm
+ * @returns void
+ */
+void findAllLCS(char *string1, char *string2)
```

Comment:
The test function should use `assert` and be named `test` (see the [typical structure](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#typical-structure-of-a-program) of a program).

Reply:
Added test function with assert